// @laser-dac/draw/dist/Rect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rect = void 0;
const Shape_1 = require("./Shape");
const Line_1 = require("./Line");
const Wait_1 = require("./Wait");
const constants_1 = require("./constants");
//
const WAIT_AMOUNT = constants_1.MAX_WAIT_AMOUNT / 2;
class Rect extends Shape_1.Shape {
    constructor(options) {
        super();
        this.x = options.x;
        this.y = options.y;
        this.width = options.width;
        this.height = options.height;
        this.color = options.color;
    }
    draw(resolution) {
        return [
            // Top.
            ...new Line_1.Line({
                from: { x: this.x, y: this.y },
                to: { x: this.x + this.width, y: this.y },
                color: this.color,
                blankBefore: true,
            }).draw(resolution),
            ...new Wait_1.Wait({
                x: this.x + this.width,
                y: this.y,
                color: this.color,
                amount: WAIT_AMOUNT,
            }).draw(),
            // Right.
            ...new Line_1.Line({
                from: { x: this.x + this.width, y: this.y },
                to: { x: this.x + this.width, y: this.y + this.height },
                color: this.color,
            }).draw(resolution),
            ...new Wait_1.Wait({
                x: this.x + this.width,
                y: this.y + this.height,
                color: this.color,
                amount: WAIT_AMOUNT,
            }).draw(),
            // Bottom.
            ...new Line_1.Line({
                from: { x: this.x + this.width, y: this.y + this.height },
                to: { x: this.x, y: this.y + this.height },
                color: this.color,
            }).draw(resolution),
            ...new Wait_1.Wait({
                x: this.x,
                y: this.y + this.height,
                color: this.color,
                amount: WAIT_AMOUNT,
            }).draw(),
            // Left.
            ...new Line_1.Line({
                from: { x: this.x, y: this.y + this.height },
                to: { x: this.x, y: this.y },
                color: this.color,
                blankAfter: true,
            }).draw(resolution),
        ];
    }
}
exports.Rect = Rect;
