// @laser-dac/draw/dist/Scene.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scene = void 0;
// Frames per second; 30fps will be enough for most use-cases.
const DEFAULT_FPS = 30;
const DEFAULT_RESOLUTION = 500;
class Scene {
    constructor(options) {
        this.points = [];
        this.resolution = (options && options.resolution) || DEFAULT_RESOLUTION;
    }
    add(shape, transformer) {
        let points = shape.draw(this.resolution);
        if (transformer) {
            points = transformer(points);
        }
        this.points = this.points.concat(points);
    }
    reset() {
        this.points = [];
    }
    start(renderFrame, fps = DEFAULT_FPS) {
        const ms = 1000 / fps;
        this.interval = setInterval(() => {
            this.reset();
            renderFrame();
        }, ms);
    }
    stop() {
        this.pause();
        this.reset();
    }
    pause() {
        if (this.interval) {
            clearInterval(this.interval);
            this.interval = undefined;
        }
    }
}
exports.Scene = Scene;
