// @laser-dac/draw/dist/Timeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Timeline = void 0;
const Shape_1 = require("./Shape");
class Timeline extends Shape_1.Shape {
    constructor(options) {
        super();
        this.items = options.items;
        this.loop = options.loop || false;
        this.started = Date.now();
    }
    reset() {
        this.started = Date.now();
    }
    getItem() {
        const now = Date.now();
        const diffInMs = now - this.started;
        let currentMs = 0;
        const currentItem = this.items.find((item) => {
            const oldMs = currentMs;
            const newMs = (currentMs += item.duration);
            return diffInMs >= oldMs && diffInMs <= newMs;
        });
        if (this.loop && !currentItem) {
            this.reset();
        }
        return currentItem;
    }
    draw(resolution) {
        const item = this.getItem();
        if (item) {
            return item.render().draw(resolution);
        }
        else {
            return [];
        }
    }
}
exports.Timeline = Timeline;
