// @laser-dac/core/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DAC = exports.Device = void 0;
const DEFAULT_POINTS_RATE = 30000;
const DEFAULT_FPS = 30;
class Device {
}
exports.Device = Device;
class DAC {
    constructor() {
        this.devices = [];
    }
    use(device) {
        this.devices.push(device);
    }
    remove(device) {
        const index = this.devices.indexOf(device);
        if (index) {
            this.devices.splice(index, 1);
        }
    }
    removeAll() {
        this.devices = [];
    }
    async start() {
        for (const device of this.devices) {
            const success = await device.start();
            if (!success) {
                return false;
            }
        }
        return true;
    }
    async stop() {
        for (const device of this.devices) {
            await device.stop();
        }
    }
    stream(scene, pointsRate = DEFAULT_POINTS_RATE, fps = DEFAULT_FPS) {
        for (const device of this.devices) {
            device.stream(scene, pointsRate, fps);
        }
    }
}
exports.DAC = DAC;
